core/src/Serial/Kokkos_Serial.cpp

```diff
@@ -59,6 +68,16 @@ void SerialInternal::finalize() {
   }
 
   m_is_initialized = false;
+
+  // guard erasing from all_instances
+  {
+    std::scoped_lock lock(all_instances_mutex);
+    auto it = std::find(all_instances.begin(), all_instances.end(), this);
+    if (it == all_instances.end())
+      Kokkos::abort(
+          "Execution space instance to be removed couldn't be found!");
+    all_instances.erase(it);
```

Comment:
I don't know if we are concerned about minimizing the amount of time the lock is held.  If so,
```suggestion
    swap(*it, all_instances.back());
    all_instances.pop_back();
```
is O(1) (although his doesn't change the overall complexity of `finalize()` as `find` is still O(N)).